infrastructure-playbooks/shrink-mgr.yml

```diff
@@ -0,0 +1,131 @@
+#
+# Use it like this:
+# ansible-playbook shrink-mgr.yml -e mgr_to_kill=ceph-mgr1
+#     Prompts for confirmation to shrink, defaults to no and
+#     doesn't shrink the cluster and yes shrinks the cluster.
+#
+# ansible-playbook -e ireallymeanit=yes|no shrink-mgr.yml
+#     Overrides the prompt using -e option. Can be used in
+#     automation scripts to avoid interactive prompt.
+
+
+- name: gather facts and check the init system
+  hosts:
+    - "{{ mon_group_name | default('mons') }}"
+    - "{{ mgr_group_name | default('mgrs') }}"
+  become: true
+  tasks:
+    - debug:
+        msg: gather facts on all Ceph hosts for following reference
+
+- name: confirm if user really meant to remove manager from the ceph cluster
+  hosts: localhost
+  become: true
+  vars_prompt:
+    - name: ireallymeanit
+      prompt: Are you sure you want to shrink the cluster?
+      default: 'no'
+      private: no
+  pre_tasks:
+    - import_role:
+        name: ceph-defaults
+
+    - import_role:
+        name: ceph-facts
+
+    - name: set_fact container_exec_cmd
+      when: containerized_deployment | bool
+      set_fact:
+        container_exec_cmd: "{{ container_binary }} exec ceph-mon-{{ hostvars[groups[mon_group_name][0]]['ansible_hostname'] }}"
+
+    - name: get total number of mgrs in cluster
+      block:
+        - name: save mgr dump output
+          command: "{{ container_exec_cmd | default('') }} ceph --cluster {{cluster}} mgr dump"
+          register: mgr_dump
+          delegate_to: "{{ groups[mon_group_name][0] }}"
+
+        - name: get number of mgrs
+          set_fact:
+            standby_mgrs: "{{ (mgr_dump.stdout | from_json)['standbys'] | list | length }}"
+
+    - name: exit playbook, if there's no standby manager
+      fail:
+        msg: "You are about to shrink the only manager present in the cluster."
+      when: standby_mgrs | int < 1
+
+    - name: exit playbook, if no manager was given
+      fail:
+        msg: "mgr_to_kill must be declared
+              Exiting shrink-cluster playbook, no manager was removed.
+              On the command line when invoking the playbook, you can use
+              -e mgr_to_kill=ceph-mgr01 argument. You can only remove a single
+              manager each time the playbook runs."
+      when: mgr_to_kill is not defined
+
+    - name: exit playbook, if the manager is not part of the inventory
+      fail:
+        msg: "It seems that the host given is not part of your inventory,
+              please make sure it is."
+      when: mgr_to_kill not in groups[mgr_group_name]
+
+    - name: exit playbook, if user did not mean to shrink cluster
+      fail:
+        msg: "Exiting shrink-mgr playbook, no manager was removed.
+           To shrink the cluster, either say 'yes' on the prompt or
+           or use `-e ireallymeanit=yes` on the command line when
+           invoking the playbook"
+      when: ireallymeanit != 'yes'
+
+    - name: exit playbook, if can not connect to the cluster
```

Comment:
this task should be run before any task trying to connect to the cluster

Reply:
Done.